### ISSUE TITLE
Tensorflow 2.0 does not iterate through entire dataset when tf.keras.model.fit is called

### ISSUE THREAD

**Issue by michaelarfreed** [tensorflow/tensorflow#34585]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Mac OS 10.14.6**
- TensorFlow installed from (source or binary): **source**
- TensorFlow version (use command below): **2.0**
- Python version: **3.6.0**

**Describe the current behavior**
I am training a model in tf.keras with tensorflow 2.0. **I am having an issue where my model appears to train successfully, but it is not iterating through the entire dataset.** A few things indicate that it's not going through the entire dataset: 

1. the model is training super fast (an epoch should take ~45 minutes, which is estimated, but it always finishes in < 1 minute),
2. the validation metrics are always reported 0. and,
3. the progress bar that prints out never fills up, and it stops at ~1/1000 of the data. 

The progress bar during training looks something like this:

```
Epoch 1/300   192/162636 [..............................] - ETA: 45:42 - loss: 0.4783 - mean_squared_error: 0.4783 - val_loss: 0.0000e+00 - val_mean_squared_error: 0.0000e+00
Epoch 2/300   186/162636 [..............................] - ETA: 45:42 - loss: 0.4783 - mean_squared_error: 0.4783 - val_loss: 0.0000e+00 - val_mean_squared_error: 0.0000e+00
...
Epoch X/300   192/162636 [..............................] - ETA: 45:42 - loss: 0.4783 - mean_squared_error: 0.4783 - val_loss: 0.0000e+00 - val_mean_squared_error: 0.0000e+00
```

 I restructured the code into tensorflow 1.15, and I do not have this issue. When I call `tf.compat.v1.enable_v2_behavior()`, I see this behavior again. There are no errors, warnings, or info that is reported, it just stops iterating through my dataset early. I am following [this tutorial](https://machinelearningmastery.com/how-to-develop-lstm-models-for-time-series-forecasting/#comment-510384) for Multiple Input Series. 

**Describe the expected behavior**
The expected behavior is that an epoch will completely go through the dataset, and it will report a reasonable validation loss, and it will take some time to go through the entire dataset. I see the correct behavior in TF 1.15, and the epochs take ~45 minutes to complete (as expected), the validation metrics are calculated, and the progress bar looks something like this (which is nothing special :) )
```
Epoch 16/300
162605/162636 [============================>.] - ETA: 0s - loss: 1.2883e-05
162636/162636 [==============================] - 2946s 1ms/sample - loss: 1.2883e-05 - val_loss: 1.5680e-05
Epoch 17/300
162605/162636 [============================>.] - ETA: 0s - loss: 1.2631e-05
162636/162636 [==============================] - 2688s 5ms/sample - loss: 1.2633e-05 - val_loss: 2.1342e-05
```

**Code to reproduce the issue**
I have a time-series dataset. It is very small so I am able to load it into memory, so I do not need the dataset API. I am windowing the time-series to produce two arrays, X and Y, and it looks something like this, 

```
X=[
   [[1,2,3],[4,5,6],   [7,8,9]],
   [[4,5,6],[7,8,9],   [10,11,12]],
   [[7,8,9],[10,11,12],[13,14,15]],
   ...
  ] 
Y = [
     [4],
     [7],
     [10],
     ...
    ]
```

(yes, I realize that I could just as easily only include one of the features. I've tried that, i.e. `X=[[[1,2,3]], [[4,5,6]], [[7,8,9]], ...], and it still doesn't work)

Then, I build my model:

```
model = Sequential()
model.add(LSTM(50, activation='relu', input_shape=(n_steps, n_features)))
model.add(Dense(1))
model.compile(optimizer='adam', loss='mse')
```

and then I train it:

```
model.fit([X],[Y],num_epochs=300,validation_split=0.2)
```

It correctly reports the number of train and validation samples, and then the progress bar pops up... but that's where the success stops. The val_loss and val_mean_squared_error is always 0, for every epoch, and it appears to never train more than a fraction (~1/1000) of my dataset, although that fraction varies slightly between epochs. This is the print out:

```
Epoch 1/300   192/162636 [..............................] - ETA: 45:42 - loss: 0.4783 - mean_squared_error: 0.4783 - val_loss: 0.0000e+00 - val_mean_squared_error: 0.0000e+00
Epoch 2/300   186/162636 [..............................] - ETA: 45:42 - loss: 0.4783 - mean_squared_error: 0.4783 - val_loss: 0.0000e+00 - val_mean_squared_error: 0.0000e+00
...
Epoch X/300   192/162636 [..............................] - ETA: 45:42 - loss: 0.4783 - mean_squared_error: 0.4783 - val_loss: 0.0000e+00 - val_mean_squared_error: 0.0000e+00
```

When I call `tf.compat.v1.enable_v2_behavior()` in TF 1.15, the behavior is the same as TF 2.0.

**Other info / logs**
[Here is a link to the StackOverflow question that I proposed before I had confirmed that this is a TensorFlow bug](https://stackoverflow.com/questions/58826512/tensorflow-2-0-does-not-iterate-through-entire-dataset-when-tf-keras-model-fit-i/59001781#59001781)

**Comment by ravikyram** [comment]
@michaelarfreed 

Looks like code is incomplete.Please, help us with simple stand alone code to reproduce the issue in our environment. It helps us in localizing the issue faster. Thanks!

**Comment by michaelarfreed** [comment]
here's the full code, with a simplified example dataset:

```
def main():
    # tf.compat.v1.enable_v2_behavior()

    # create the dataset
    datasetD = make_dataset()

    train_x = datasetD['train']['x']
    train_y = datasetD['train']['y']
    # test_x  = datasetD['test']['x']
    # test_y  = datasetD['test']['y']
    val_x   = datasetD['val']['x']
    val_y   = datasetD['val']['y']

    # create the model
    model = create_model()
    model.compile(optimizer='adam',
                  loss='mse')
    
    print(model.summary())

    history = model.fit([train_x],[train_y],
                        batch_size=32,
                        epochs=300, 
                        validation_data=([val_x],[val_y]),
                        validation_freq=1)

def make_dataset():
    input_window_samps  = 50
    num_signals         = 1
    output_window_samps = 3
    returnD = {}

    returnD['train'] = {}
    returnD['train']['x'] = []
    returnD['train']['y'] = []
    
    for i in range(10000):
        returnD['train']['x'].append(np.arange(i,i+input_window_samps))
        returnD['train']['y'].append(np.expand_dims(np.arange(i+input_window_samps,i+input_window_samps+output_window_samps),axis=1))
    
    returnD['val'] = {}
    returnD['val']['x'] = []
    returnD['val']['y'] = []
    
    for i in range(10000,20000):
        returnD['val']['x'].append(np.arange(i,i+input_window_samps))
        returnD['val']['y'].append(np.expand_dims(np.arange(i+input_window_samps,i+input_window_samps+output_window_samps),axis=1))
    
    returnD['test'] = {}
    returnD['test']['x'] = []
    returnD['test']['y'] = []
    
    for i in range(20000,30000):
        returnD['test']['x'].append(np.arange(i,i+input_window_samps))
        returnD['test']['y'].append(np.expand_dims(np.arange(i+input_window_samps,i+input_window_samps+output_window_samps),axis=1))

    return returnD

def create_model():
    # from https://machinelearningmastery.com/how-to-develop-lstm-models-for-time-series-forecasting/ "Multiple Parallel Input and Multi-Step Output" example 
    input_window_samps  = 50
    num_signals         = 1
    output_window_samps = 3
    units0 = 10
    units1 = 10

    input = Input(shape=(input_window_samps*num_signals,))
    x = Reshape((input_window_samps,num_signals))(input)
    x = LSTM(units0,activation='relu')(x)
    x = RepeatVector(output_window_samps)(x)
    x = LSTM(units1,activation='relu',return_sequences=True)(x)
    x = TimeDistributed(Dense(num_signals))(x)

    model = Model(inputs=input,outputs=x)
    return model
    ```

**Comment by ravikyram** [comment]
@michaelarfreed 

I have tried on colab with TF version 1.15 and i do not see any issue.But with TF 2.0 i am able to reproduce the issue. However it looks like it is resolved with TF nightly version (`!pip install tf-nightly==2.1.0dev20191203`). Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/e751dbf44cf2051a5117caaf663ccb94/untitled447.ipynb).Thanks!

**Comment by michaelarfreed** [comment]
Thank you! I only see the error on TF 1.15 when I uncomment the line `tf.compat.v1.enable_v2_behavior()`. When it is in TF 1.15, I do not see any issue, either. Sorry for the confusion about that.
I see that the epoch reports the validation metrics in the nightly build, but I noticed that the progress bar still does not completely fill up and each epoch takes significantly shorter time than it did in TF 1.15. I just want to confirm that the behavior I'm seeing on my machine matches your behavior, and that every batch is being processed. This is the output I'm getting from the training:
```
Epoch 1/300
   64/10000 [..............................] - ETA: 8:19 - loss: 31558043.8788 - val_loss: 234850827.5200Epoch 2/300
   64/10000 [..............................] - ETA: 3:31 - loss: 39555757.7273 - val_loss: 234850769.5104Epoch 3/300
   64/10000 [..............................] - ETA: 3:28 - loss: 38133528.6667 - val_loss: 234850704.8192Epoch 4/300
   64/10000 [..............................] - ETA: 3:25 - loss: 32333142.6818 - val_loss: 234850648.7808Epoch 5/300
   64/10000 [..............................] - ETA: 3:30 - loss: 25878961.6970 - val_loss: 234850591.9232Epoch 6/300
```

**Comment by ravikyram** [comment]
I am able to see the issue with TF 1.15 when we uncomment the line `tf.compat.v1.enable_v2_behavior()`. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/7e597c68b2d988befbc9cd6f42c2a55a/untitled478.ipynb).Thanks!

**Comment by michaelarfreed** [comment]
Thank you, @ravikyram !

**Comment by michaelarfreed** [comment]
Were you able to look into the other issue that I noticed in TF 2.0, where the ETA reported for an epoch is much higher than the amount of time an epoch takes? In TF 1.15 with `tf.compat.v1.enable_v2_behavior()` commented out, it correctly estimates the ETA and an epoch takes the amount of time listed by ETA (+/- a few seconds).  In TF 2.0 (in both the nightly build and the release) and in TF 1.15 with `tf.compat.v1.enable_v2_behavior()` uncommented, it estimates the same ETA as in TF 1.15 with `tf.compat.v1.enable_v2_behavior()` commented out, but an epoch is over within <30 seconds, even when eta > 2 minutes.

(From my previous comment)
> I see that the epoch reports the validation metrics in the nightly build, but I noticed that the progress bar still does not completely fill up and each epoch takes significantly shorter time than it did in TF 1.15. I just want to confirm that the behavior I'm seeing on my machine matches your behavior, and that every batch is being processed.

**Comment by jsimsa** [comment]
@rachellim could you please take a look? thanks

**Comment by michaelarfreed** [comment]
To give a more concrete example and expand on my previous comment, when I use my true dataset, in TF 1.15 when `tf.compat.v1.enable_v2_behavior()` is commented out, the ETA for an epoch is ~30 minutes, and an epoch takes around 30 minutes. In TF 2.0 and in TF 1.15 with `tf.compat.v1.enable_v2_behavior()` uncommented, the ETA for each epoch is still 30 minutes, but each epoch takes < 30 seconds, and this speed-up seems too good to be true. I also see this speedup in the nightly build that is in your notebook. The effect is less noticeable in the example dataset that I provided since the ETA is low, but it is there.

**Comment by michaelarfreed** [comment]
I noticed that the bug that I have described is also present in the TF 2.1 release. If I call `!pip install tensorflow==2.1` where you have called `!pip install tensorflow==2.0` in [this notebook](https://colab.sandbox.google.com/gist/ravikyram/e751dbf44cf2051a5117caaf663ccb94/untitled447.ipynb), I see the same behavior that I have been describing (1. the progress bar does not fill up for a full epoch, 2. the ETA for an epoch is 4+ minutes, but an epoch finishes in seconds). The one thing that is fixed in the 2.1 release is that val_loss is calculated.

Please let me know if you need any more information from me to resolve this bug, I really appreciate your help and work!!

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF 2.6.0-dev20210529,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/df977b60c021ddb6a0d58d33ba4963cb/untitled91.ipynb#scrollTo=hP3UL5gK9TgP).. Thanks !
